doc/neps/nep-0047-array-api-standard.rst

```diff
@@ -0,0 +1,534 @@
+
+The most obvious omissions are the complex dtypes. The rationale for the lack
+of complex support in the first version of the array API standard is that several
+libraries (PyTorch, MXNet) are still in the process of adding support for
+complex dtypes. The next version of the standard is expected to include ``complex32``
+and ``complex64`` (see `this issue <https://github.com/data-apis/array-api/issues/102>`__
+for more details).
+
+Specifying dtypes to functions, e.g. via the ``dtype=`` keyword, is expected
+to only use the dtype literals. Format strings, Python builtin dtypes, or
+string representations of the dtype literals are not accepted - this will
+improve readability and portability of code at little cost.
+
+Casting rules are only defined between different dtypes of the same kind. The
+rationale for this is that mixed-kind (e.g., integer to floating-point)
+casting behavior differs between libraries. NumPy's mixed-kind casting
+behavior doesn't need to be changed or restricted, it only needs to be
+documented that if users use mixed-kind casting, their code may not be
+portable.
+
+.. image:: _static/nep-0047-casting-rules-lattice.png
+
+*Type promotion diagram. Promotion between any two types is given by their
+join on this lattice. Only the types of participating arrays matter, not
+their values. Dashed lines indicate that behaviour for Python scalars is
+undefined on overflow. Boolean, integer and floating-point dtypes are not
+connected, indicating mixed-kind promotion is undefined.*
+
+The most important difference between the casting rules in NumPy and in the
+array API standard is how scalars and 0-dimensional arrays are handled. In
+the standard, array scalars do not exist and 0-dimensional arrays follow the
+same casting rules as higher-dimensional arrays.
+
+See the `Type Promotion Rules section of the array API standard <https://data-apis.github.io/array-api/latest/API_specification/type_promotion.html>`__
+for more details.
+
+.. note::
+
+    It is not clear what the best way is to support the different casting rules
+    for 0-dimensional arrays and no value-based casting. One option may be to
+    implement this second set of casting rules, keep them private, mark the
+    array API functions with a private attribute that says they adhere to
+    these different rules, and let the casting machinery check whether for
+    that attribute.
+
+    This needs discussion.
+
+
+Indexing
+~~~~~~~~
+
+An indexing expression that would return a scalar with ``ndarray``, e.g.
+``arr_2d[0, 0]``, will return a 0-D array with the new array object. There are
+several reasons for that: array scalars are largely considered a design mistake
+which no other array library copied; it works better for non-CPU libraries
+(typically arrays can live on the device, scalars live on the host); and it's
+simply a consistent design. To get a Python scalar out of a 0-D array, one can
+simply use the builtin for the type, e.g. ``float(arr_0d)``.
```

Comment:
A likely complaint would be that there is no way to get a Python scalar of an _unknown_ type out.

In NumPy, the idiomatic way to do this is either `arr_0d.item()` or `arr_0d[()]`, but neither is supported in the API standard.

Reply:
Good point. `[()]` wasn't thought about, should be added if all libraries support it. Most libraries have `.item()`, only tensorflow seems to use `.eval()`. A bit friendlier than empty-tuple indexing, so could add that as well.

Reply:
Actually the TF part was wrong - my TensorFlow is rusty. TF does not seem to have a good way of doing this, there's no method and indexing with an empty tuple is a no-op:
```
>>> a = tf.constant(1)
>>> type(a)
<class 'tensorflow.python.framework.ops.EagerTensor'>
>>> type(a[()])
<class 'tensorflow.python.framework.ops.EagerTensor'>
>>> b = np.array(1)
>>> type(b)
<class 'numpy.ndarray'>
>>> type(b[()])
<class 'numpy.int64'>
```

PyTorch works like TF for indexing. That all makes sense, because numpy produces array scalars, which don't exist in other libraries. And indexing shouldn't produce a Python int/float here.

PyTorch does have `.item()` which produces a Python scalar. But that's then still inconsistent with NumPy - NumPy itself arguably has no way of producing a Python scalar of unknown type.

I'm not sure if implementing `.item()` is desired or not.

Reply:
I don't think `[()]` should return a scalar. It only works that way in NumPy because array indexing returns scalars.

> But that's then still inconsistent with NumPy - NumPy itself arguably has no way of producing a Python scalar of unknown type.

Doesn't NumPy have `.item()`?

That said, maybe we don't really need `.item()` until we hear complaints about it. Unboxing is only a valid thing for some array types, and in dask.array, I believe `.item()` can actually return a `dask.delayed` object.

Reply:
What is the purpose of even mentioning Python scalars here? It seems irrelevant to the question of indexing. Note that `arr[()]` in NumPy returns a NumPy scalar, not a Python scalar, even if the array is 0-dimensional:

```py
>>> type(np.array(1.0)[()])
<class 'numpy.float64'>
```

It might be clearer to just have a separate section that makes it clear that there are no scalar types in the array API, only 0-d arrays. That can be mentioned here in the indexing section, but I think it's about more than just indexing (indexing is just one way that you can end up with a scalar, though generally the most common). Right now this is sort of mentioned in a few places, but it's important enough to warrant a separate section.
```suggestion
simply a consistent design. To get a Python scalar out of a 0-D array, one can
simply use the builtin for the type, e.g. ``float(arr_0d)``.
```
```suggestion
simply a consistent design. To get a Python scalar out of a 0-D array, one can
simply use the builtin for the type, e.g. ``float(arr_0d)``.
```

And even then, *Python scalars* are a whole other thing. Mixing them up in the same section only confuses things, especially if we use the same terminology ("scalar") for both. The spec can probably get away with this (though IMO it should also be less ambiguous), because it doesn't have scalar types, but NumPy does, so it would help to be clear about these points and to keep them separate.

Reply:
@asmeurer good question, I'll see if a separate section makes sense. The array scalars point itself is minor, and no one will be surprised that they're excluded. The "get me a Python scalar" thing is more important, and there should be a good way of getting one.

Reply:
The outcome of this discussion in the meeting was that adding empty tuple indexing would be good for consistency, even though it's a no-op.

I think there's nothing else to do here. You cannot get a Python scalar of unknown type with NumPy today, so no urgent need to invent a new way of doing that.